gluetool/action.py

```diff
@@ -0,0 +1,316 @@
+
+    def close(self):
+        # type: () -> None
+
+        if not Tracer.TRACER:
+            return
+
+        from .utils import wait
+
+        # yield to IOLoop to flush the spans - https://github.com/jaegertracing/jaeger-client-python/issues/50
+        time.sleep(2)
+
+        future = Tracer.TRACER.close()
+
+        wait('tracing flush', future.done, timeout=self.flush_timeout, tick=2, logger=self.logger)
+
+
+class Action(object):
+    """
+    A piece of a workflow: it has a name, and starts and ends at some point of time. Represents an individual unit
+    of work.
+
+    :param str label: a human-readable string which concisely represents the work done by the ``Action``. The name
+        should be the most general string that describes an interesting class of ``Action`` instances.
+        I.e. ``fetch-url` is better than ``fetch-url-https://foo.com```.
+    :param Action parent: parent ``Action`` - one action can spawn multiple additional "children" actions to
+        achieve its goal, either explicitly or by using instrumented library code.
+    :param dict tags: additional key/value tags of this action, e.g. ``url=https://foo.com``.
+    :param ContextAdapter logger: logger to use for logging purposes.
+    """
+
+    # For each thread, we keep a LIFO of unfinished actions. The topmost one is considered to be
+    # "current".
+    #
+    # Works as long as user keeps only a single "active" action in a thread - if one creates two actions,
+    # side by side, the last one becomes the "current":
+    #
+    # A1, A2 = Action(), Action()
+    #     ^ current action
+    _thread_actions = threading.local()
+
+    @staticmethod
+    def _action_stack():
+        # type: () -> List[Action]
+        """
+        Return current - or create an empty new one - list of unfinished actions of the current thread.
+        """
+
+        if not hasattr(Action._thread_actions, 'stack'):
+            Action._thread_actions.stack = []
+
+        return cast(
+            List[Action],
+            Action._thread_actions.stack
+        )
+
+    @staticmethod
+    def _add_action(action):
+        # type: (Action) -> None
+        """
+        Add action on top of the list of unfinished actions of the current thread.
+        """
+
+        Action._action_stack().append(action)
+
+    @staticmethod
+    def _drop_action(action):
+        # type: (Action) -> None
+        """
+        Drop action from the list of unfinished actions of the current thread.
+        """
+
+        Action._action_stack().remove(action)
+
+    @staticmethod
+    def current_action():
+        # type: () -> Action
+        """
+        Return the top-most - "current" - unfinished action of the current thread.
+        """
+
+        stack = Action._action_stack()
+
+        if not stack:
+            raise RuntimeError('Action stack is empty')
+
+        return stack[-1]
+
+    @staticmethod
+    def set_thread_root(action):
```

Comment:
What difference between this method and `_add_action` called with the empty stack? I just curious about the purpose of the method.

Reply:
I extended the docstring. `_add_action` does similar thing, but adds action to the stack, not re-setting it, while this method is intended to be used whenever someone starts some work inside a - possibly reused - worker thread, re-setting the stack for the thread before doing anything else. It could be written as:

```python
Action._action_stack()[:] = []
Action._add_action(action)
```

Reply:
I see. Thank you!